Removed some node references to prevent memory leaks

kukulich opened the pull request:
Another part of https://github.com/Roave/BetterReflection/issues/1073

kukulich committed:
Removed ReflectionEnumCase::getAst() to prevent memory leaks

kukulich committed:
Removed ReflectionConstant::getAst() to prevent memory leaks

kukulich committed:
Removed $node property in ReflectionAttribute to prevent memory leaks

Ocramius commented on src/Reflection/ReflectionAttribute.php:
```diff
@@ -11,19 +11,31 @@
+
 class ReflectionAttribute
 {
+    private string $name;
```
Should we have `non-empty-string` on this?

kukulich replied:
Done.

Ocramius commented on src/Reflection/ReflectionAttribute.php:
```diff
@@ -11,19 +11,31 @@
 {
+    private string $name;
+
+    /** @var array<int|string, Node\Expr> */
```
Is mixing `int` and `string` arguments allowed?

Wondering if this could be `array<int, Node\Expr>|array<string, Node\Expr>` :thinking:

kukulich replied:
I think `array<int|string, Node\Expr>` is the most accurate for code like this:

```
#[Attr('arg1', 'arg2', arg3: self::class, arg4: [0, 1]])]
```

Ocramius replied:
Works for me :+1:

Ocramius commented on src/Reflection/ReflectionAttribute.php:
```diff
@@ -11,19 +11,31 @@
+        }
     }
 
     public function getName(): string
```
Can we somehow declare `non-empty-string` from here?

kukulich replied:
Done.

Ocramius commented on src/Reflection/ReflectionConstant.php:
```diff
@@ -25,19 +25,50 @@
+
+    private int $startColumn;
+
+    private int $endColumn;
```
We should declare types like `non-empty-string` and `positive-int` where possible, on these

kukulich replied:
Done.

Ocramius commented on src/Reflection/ReflectionEnumCase.php:
```diff
@@ -22,13 +23,38 @@
 
 class ReflectionEnumCase
 {
+    private string $name;
+
+    private Node\Expr|null $value;
+
+    /** @var list<ReflectionAttribute> */
+    private array $attributes;
+
+    private string $docComment;
+
+    private int $startLine;
+
+    private int $endLine;
+
+    private int $startColumn;
+
+    private int $endColumn;
```
We should declare stricter types like `positive-int` or `non-empty-string` for these

kukulich replied:
Done.

kukulich committed:
More strict types in ReflectionParameter

kukulich committed:
Added ReflectionParameter::getStartLine() and getEndLine()

kukulich commented:
I've also added strict types to `ReflectionParameter` (properties added in #1213) and added `getStartLine()` and `getEndLine()` methods - I really don't know why they were missing 😄

Ocramius commented on src/Reflection/ReflectionEnumCase.php:
```diff
@@ -22,13 +23,52 @@
+        EnumCase $node,
         private ReflectionEnum $enum,
     ) {
+        /** @var non-empty-string $name */
```
Hmm, can our CS tooling replace some of these with assertions?

Good enough for now, just don't like having a leap of faith

kukulich replied:
Probably in the next release, I know the author 😄

Ocramius reviewed (approved):
Most excellent, thanks @kukulich!
